lib/offline.js

```diff
@@ -55,60 +61,121 @@ module.exports = function(config) {
       } catch (ex) {
         return '';
       }
-    }).then(function(cacheId) {
-      var staticFileGlobs = fileGlobs.map(function(v) {
+    })
+    .then(function(cacheId) {
+      checkImports(importScripts, rootDir);
+
+      var absoluteGlobs = fileGlobs.map(function (v) {
         return path.join(rootDir, v);
       });
+      var files = flatGlobs(absoluteGlobs);
+      var filesAndHashes = getFilesAndHashes(files, rootDir, absoluteGlobs);
+
+      var replacements = {
+        cacheId: cacheId,
+        ignoreUrlParametersMatching: config.ignoreUrlParametersMatching || [/./],
+        cacheVersion: getHashOfHashes(pluckHashes(filesAndHashes)),
+        resources: filesAndHashes,
+        importScripts: importScripts,
+      };
 
-      staticFileGlobs.forEach(function(globPattern) {
-        glob.sync(globPattern.replace(path.sep, '/')).forEach(function(file) {
-          var stat = fs.statSync(file);
+      var stream = gulp.src([__dirname + '/../templates/app/offline-worker.js'])
+        .pipe(template(replacements))
+        .pipe(gulp.dest(rootDir));
 
-          if (stat.isFile() && stat.size > 2 * 1024 * 1024 && staticFileGlobs.indexOf(file) === -1) {
-            console.log(chalk.yellow.bold(file + ' is bigger than 2 MiB. Are you sure you want to cache it? To suppress this warning, explicitly include the file in the fileGlobs list.'));
-          }
-        });
+      return new Promise(function (resolve, reject) {
+        stream.on('finish', function () { resolve(); });
+        stream.on('error', function (e) { reject(e); });
       });
+    })
+    .then(resolve, reject);
 
-      var importScripts = config.importScripts || [];
-      importScripts.forEach(function(script) {
-        var stat;
-        try {
-          stat = fs.statSync(path.join(rootDir, script));
-        } catch (ex) {
-          console.log(chalk.red.bold(script + ' doesn\'t exist.'));
-          throw ex;
-        }
-
-        if (!stat.isFile()) {
-          console.log(chalk.red.bold(script + ' is not a file.'));
-          throw new Error(script + ' is not a file.');
-        }
+    function flatGlobs(fileGlobs) {
+      return Object.keys(fileGlobs.reduce(function (matches, fileGlob) {
+        fileGlob = fileGlob.replace(path.sep, '/');
+        glob.sync(fileGlob, { nodir: true }).forEach(function (filepath) {
+          matches[filepath] = filepath;
+        });
+        return matches;
+      }, {}));
+    }
+
+    function getFilesAndHashes(files, rootDir, sizeWhiteList) {
+      var totalSize = 0;
+      var filesAndHashes = files.map(function (filepath) {
+        var size = countFileSize(filepath, sizeWhiteList);
+        totalSize += size;
+        var data = fs.readFileSync(filepath);
+        var hash = getHash(data);
+        logOk(filepath, size);
+        return {
+          path: filepath.replace(rootDir, function (match, offset) {
+            //XXX: The root must be the worker's directory
```

Comment:
You could replace it with "NB" (from the Italian "nota bene", i.e. "note well") or simply "Note", but I'd still be interested in more details about what this comment is noting, as I don't quite understand the point being made.